### DISCUSSION TITLE
Display a brief list of available servers within the UI

### DISCUSSION THREAD

**Discussion by Nexxed** [alexjustesen/speedtest-tracker#25] [Ideas]
Having to manually fetch server IDs can be annoying, so a neat QoL feature would be to parse the output of the `--servers` option and display the results in the UI for us to pick from - to either blacklist or insert into the servers list.

The behaviour of the `--servers` option just lists the top ten of the nearest servers (probably determined by ping) which makes it ideal for this sort of thing :)

**Comment by Nexxed** [comment]
Aside from parsing the output of the option, you could also use the same API as the official Speedtest.net tool: https://www.speedtest.net/api/js/servers?engine=js&limit=10&https_functional=true

The limit is capped to 20 (default is 10) so it can allow for more servers for us to pick from.

An obvious downside to this is future compatibility as that API endpoint can change at any given time without notice, but I still found it to be something worth mentioning!

**Reply by alexjustesen** [reply]
I was looking at the server selector the other day and looks like we might be able to build in a little search function too and looks like it's hitting this endpoint https://www.speedtest.net/api/js/servers?engine=js&search=cal&https_functional=true&limit=100

**Reply by Nexxed** [reply]
That could be pretty powerful! We could even sort the results based on distance and display more information that may be relevant to the end-user.

**Reply by alexjustesen** [reply]
Now we're cooking with fire!

**Comment by alexdelprete** [comment]
It's a nice idea, but I hope we'll be able to manually select servers outside the top X that would come out from the automatic process. From my personal experience, the autoselected servers are often not really the best ones. After thousands of tests, I found out which servers (2-3) are best in terms of speed throughout the day, availability and other factors. But it's cool to be able to pick the servers in a managed way through the UI.

**Reply by Nexxed** [reply]
What if we could configure the tracker to run a configurable amount of extra Speedtests per "check"?

Each test would be ran against a different provider/server and we could compare all of the results and use the highest-achieving one as the final result?

This could reduce the chances of landing on a "bad" server and can help to display a true value that isn't determined by server congestion 😄

**Reply by alexdelprete** [reply]
User option:

- **batch-test**: provide results for each server tested, with stats (avg./top/worst for each metric: DL/UL/PING/etc.)
- **round-robin**: test one server at a time from the list, with stats
- other ideas? :)

**Reply by Nexxed** [reply]
Those options are exactly what I was thinking!

There was this idea that I had a while back of running traceroutes to specified hosts at regular intervals to track routing and latency changes between you and other providers (such as AWS, GCP, CloudFlare, and more).

This seems a bit out of scope for this project (as it's for speed testing) but it is somewhat related, so I thought I'd mention it anyway in case it's something you find worth adding to the project :)

Edit: This was a quick afterthought, but maybe some in-depth statistics for buffer bloat would be helpful? Not sure what those would be though.

**Reply by alexdelprete** [reply]
It's not out of scope...we could extend the project to latency metrics, speedtest-cli provides the necessary data:

```shell
root@rpi4time:~ # speedtest -s 3301

   Speedtest by Ookla

      Server: Connesi s.p.a. - Foligno (id: 3301)
         ISP: Tecno General S.r.l
Idle Latency:    17.27 ms   (jitter: 0.09ms, low: 17.16ms, high: 17.32ms)
    Download:   889.56 Mbps (data used: 1.0 GB)
                128.06 ms   (jitter: 53.73ms, low: 17.47ms, high: 2125.86ms)
      Upload:   214.45 Mbps (data used: 388.5 MB)
                 17.61 ms   (jitter: 0.36ms, low: 17.18ms, high: 21.64ms)
 Packet Loss:     0.0%
  Result URL: https://www.speedtest.net/result/c/4ca862e9-155d-4af5-aea1-c28d41b87f70

root@rpi4time:~ # speedtest -s 4302

   Speedtest by Ookla

      Server: Vodafone IT - Milan (id: 4302)
         ISP: Tecno General S.r.l
Idle Latency:     8.46 ms   (jitter: 0.27ms, low: 8.25ms, high: 8.80ms)
    Download:   849.61 Mbps (data used: 1.1 GB)
                 53.69 ms   (jitter: 44.68ms, low: 8.80ms, high: 421.38ms)
      Upload:   273.99 Mbps (data used: 338.0 MB)
                  9.17 ms   (jitter: 0.66ms, low: 8.23ms, high: 27.14ms)
 Packet Loss:     0.0%
  Result URL: https://www.speedtest.net/result/c/9af06b36-4933-4fc8-bbdc-2c8b516ee8ec
```

We could also think outside the box, for the future, about using other tools too, to get loaded/unloaded measurement to test bufferbloat etc.:

- https://fast.com
- http://www.dslreports.com/speedtest
- https://www.waveform.com/tools/bufferbloat

Maybe we could start thinking of a new name for the project...:)

**Reply by Nexxed** [reply]
There's some very interesting untapped data here that we could use to provide more insight.
What do you think? @alexjustesen 😄

**Reply by spcqike** [reply]
> User option:
> 
> * **batch-test**: provide results for each server tested, with stats (avg./top/worst for each metric: DL/UL/PING/etc.)
> * **round-robin**: test one server at a time from the list, with stats
> * other ideas? :)

User threshold to repeat the test if download / upload is < 75% of max? Value to be adjusted by user

**Reply by alexdelprete** [reply]
> User threshold to repeat the test if download / upload is < 75% of max? Value to be adjusted by user

Yes, but I see that as a general policy option, for all tests.

**Comment by alexdelprete** [comment]
fast.com provides loaded/unloaded data...but it never tops my fiber speed (1Gbps/300Mbps). I don't know if they provide an API or a CLI tool...

![image](https://user-images.githubusercontent.com/7027842/191171730-c6e9653f-5aba-4657-802a-400e860c3a8e.png)

**Reply by Nexxed** [reply]
Yeah, Fast.com is usually slower than Speedtest based on past experience 😕

**Reply by alexjustesen** [reply]
That's interesting fast.com gives me the full speed when I test, (1Gbps/1Gbps fiber) 
![image](https://user-images.githubusercontent.com/1144087/191255913-b24e7d32-1ade-4fcf-a118-dd9a88946563.png)

**Reply by alexdelprete** [reply]
it depends on the server it uses...netflix server here in Milan is probably throttled. When I ran the test it said it used a server in Milan (IT) and one in USA. :)

BTW: can we use node.js projects in Laravel framework? This looks interesting: https://github.com/karelkryda/universal-speedtest

I also found some unofficial fast.com API project, but many people say they're not legal because fast.com uses a private API by Netflix and you're not allowed to use it...sounds strange to me...it would be cool if we could use fast.com, speedtest.net and cloudflare speedtest. :)

UPDATE: this for fast.com looks interesting https://github.com/sindresorhus/fast-cli it also supports bufferbloat, can you check it on your line?

![image](https://user-images.githubusercontent.com/7027842/191282924-38a09b36-7b39-482f-af8c-868af757a559.png)

**Reply by alexjustesen** [reply]
I do plan on supporting other test methods in the future, especially when it comes to setting up ping to various services.

I use Ubiquiti Unifi at home and I like how the controller has my speedtest results and ping to popular domains right on the dashboard 
![image](https://user-images.githubusercontent.com/1144087/191294974-884e77f3-f3e5-48c8-b74f-c266ccfa821b.png)

**Reply by alexdelprete** [reply]
This is great to hear from you. It'll be an awesome project.

**Comment by alexdelprete** [comment]
The one I like the most, for the UI, is the cloudflare [speedtest](https://speed.cloudflare.com/), we could refer to this for the results presentation:

![image](https://user-images.githubusercontent.com/7027842/191173945-8742163e-b26c-4645-b5bf-f0826b4440bf.png)

**Comment by alexjustesen** [comment]
Implemented in a past version 👍
